planner/cascades/transformation_rules.go

```diff
@@ -323,3 +327,94 @@ func (r *PushSelDownProjection) OnTransform(old *memo.ExprIter) (newExprs []*mem
 	newTopSelExpr.SetChildren(newProjGroup)
 	return []*memo.GroupExpr{newTopSelExpr}, true, false, nil
 }
+
+// PushSelDownAggregation pushes Selection down to the child of Aggregation.
+type PushSelDownAggregation struct {
+}
+
+// GetPattern implements Transformation interface.
+// The pattern of this rule is `Selection -> Aggregation`.
+func (r *PushSelDownAggregation) GetPattern() *memo.Pattern {
+	return memo.BuildPattern(
+		memo.OperandSelection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandAggregation, memo.EngineTiDBOnly),
+	)
+}
+
+// Match implements Transformation interface.
+func (r *PushSelDownAggregation) Match(expr *memo.ExprIter) bool {
+	// The aggregation must be a complete mode.
+	agg := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	for _, aggFunc := range agg.AggFuncs {
+		if aggFunc.Mode != aggregation.CompleteMode {
+			return false
+		}
+	}
+	return true
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `sel->agg->x` to `agg->sel->x` or `sel->agg->sel->x`
+// or just keep the selection unchanged.
+func (r *PushSelDownAggregation) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	sel := old.GetExpr().ExprNode.(*plannercore.LogicalSelection)
+	agg := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	var pushedExprs []expression.Expression
+	var remainedExprs []expression.Expression
+	exprsOriginal := make([]expression.Expression, 0, len(agg.AggFuncs))
+	for _, fun := range agg.AggFuncs {
```

Comment:
s/fun/func

Reply:
func is a keyword, I have changed it to `aggFunc`.